## Business Problem

American Mortgage Bank grapples with the complexities of an evolving real estate landscape, navigating through shifting market dynamics and fluctuating trends. As the bank aims to refine its lending strategies, enhance risk assessments, and innovate its suite of mortgage products, it encounters obstacles in deciphering the nuances of the ever-changing market.

In response to these challenges, American Mortgage Bank seeks to harness the power of data-driven insights and advanced analytics to gain a comprehensive understanding of market trends. By leveraging predictive tools and machine learning algorithms, VitaGroup aims to revolutionize the real estate market by providing financial institutions with a deeper understanding of market trends through time series analysis. This insight will inform lending strategies and facilitate innovation in the range of mortgage products offered by the bank.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv("Neighborhood_month.csv")
df.head()

RegionID  SizeRank         RegionName    RegionType StateName State  \
0    112345         0           Maryvale  neighborhood        AZ    AZ   
1    192689         1           Paradise  neighborhood        NV    NV   
2    270958         2    Upper West Side  neighborhood        NY    NY   
3    270957         3    Upper East Side  neighborhood        NY    NY   
4    118208         4  South Los Angeles  neighborhood        CA    CA   

          City                                  Metro          CountyName  \
0      Phoenix              Phoenix-Mesa-Chandler, AZ     Maricopa County   
1    Las Vegas       Las Vegas-Henderson-Paradise, NV        Clark County   
2     New York  New York-Newark-Jersey City, NY-NJ-PA     New York County   
3     New York  New York-Newark-Jersey City, NY-NJ-PA     New York County   
4  Los Angeles     Los Angeles-Long Beach-Anaheim, CA  Los Angeles County   

      2000-01-31  ...    2023-05-31    2023-06-30    2023-07-31    2023-08-31  \
0   67383.031641  ...  3.176412e+05  3.194960e+05  3.219764e+05  3.249856e+05   
1  133640.057992  ...  3.607395e+05  3.614624e+05  3.632721e+05  3.661695e+05   
2  394692.554604  ...  1.293743e+06  1.287903e+06  1.281592e+06  1.271799e+06   
3  641748.074378  ...  1.265151e+06  1.259556e+06  1.255193e+06  1.250717e+06   
4  130878.148719  ...  6.354036e+05  6.391914e+05  6.465671e+05  6.564532e+05   

     2023-09-30    2023-10-31    2023-11-30    2023-12-31    2024-01-31  \
0  3.276488e+05  3.300777e+05  3.321366e+05  3.337132e+05  3.347330e+05   
1  3.690387e+05  3.715277e+05  3.736861e+05  3.757784e+05  3.776834e+05   
2  1.261754e+06  1.250664e+06  1.238787e+06  1.231254e+06  1.225646e+06   
3  1.246179e+06  1.237942e+06  1.226768e+06  1.216496e+06  1.209650e+06   
4  6.664609e+05  6.749575e+05  6.815551e+05  6.858569e+05  6.835768e+05   

     2024-02-29  
0  3.357045e+05  
1  3.795126e+05  
2  1.219479e+06  
3  1.203629e+06  
4  6.770799e+05  

[5 rows x 299 columns]

In [3]:
#getting data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21662 entries, 0 to 21661
Columns: 299 entries, RegionID to 2024-02-29
dtypes: float64(290), int64(2), object(7)
memory usage: 49.4+ MB


In [4]:
#checking shape for the dataset
df.shape

(21662, 299)

## Data Cleaning

To ensure the development of a robust and accurate model, several data cleaning techniques will be applied to the dataset. The following techniques will be employed:

- Completeness: This technique involves addressing missing values within the dataset. Steps will be taken to identify and handle missing data appropriately, either through imputation or removal, to ensure that the dataset is complete.

- Consistency: The consistency of the data will be examined to identify any discrepancies or irregularities. Inconsistencies in variables, such as conflicting formats or conflicting information within the dataset, will be addressed and resolved to maintain data integrity.

- Validity: Validity refers to the accuracy and relevance of the data. Data validation techniques will be applied to verify that the values within each variable align with expected ranges or predefined criteria. Any invalid or erroneous data points will be rectified or removed from the dataset.

- Uniformity: Uniformity is crucial during the data cleaning process to ensure consistency and accurate analysis. Inconsistent or non-uniform data can introduce errors and bias into the modeling process, leading to unreliable results. Robust techniques will be employed to detect and handle non-uniformity effectively, ensuring that the data is standardized and aligned.

By applying these data cleaning techniques, the dataset will be refined and prepared, ensuring the reliability and accuracy of the data before proceeding with the subsequent stages of analysis and modeling.

### Completeness
To achieve completeness in our data, I will be checking for missing values in the data.

In [5]:
# Check for null values
print(f'The data has {df.isna().sum().sum()} missing values')

The data has 1505110 missing values


In [6]:
# Define a function to explore missing data
def missing_data(df):
    missing_data = df.isna().sum()
    missing_data = missing_data[missing_data>0]
    return missing_data.to_frame()

In [7]:
# expanding the number of visible columns
pd.set_option('display.max_columns', None)

In [8]:
# Apply missing_data function to the dataframe
missing_data(df).T

City  Metro  2000-01-31  2000-02-29  2000-03-31  2000-04-30  2000-05-31  \
0     2     14       10478       10400       10389       10380       10316   

   2000-06-30  2000-07-31  2000-08-31  2000-09-30  2000-10-31  2000-11-30  \
0       10303       10294       10270       10261       10252       10247   

   2000-12-31  2001-01-31  2001-02-28  2001-03-31  2001-04-30  2001-05-31  \
0       10211       10007        9943        9831        9804        9797   

   2001-06-30  2001-07-31  2001-08-31  2001-09-30  2001-10-31  2001-11-30  \
0        9787        9774        9748        9696        9660        9644   

   2001-12-31  2002-01-31  2002-02-28  2002-03-31  2002-04-30  2002-05-31  \
0        9634        9516        9507        9500        9327        9275   

   2002-06-30  2002-07-31  2002-08-31  2002-09-30  2002-10-31  2002-11-30  \
0        9093        9077        9061        9045        9030        9020   

   2002-12-31  2003-01-31  2003-02-28  2003-03-31  2003-04-30  2003-05-31  \
0        9015        8837        8827        8811        8797        8756   

   2003-06-30  2003-07-31  2003-08-31  2003-09-30  2003-10-31  2003-11-30  \
0        8748        8686        8485        8442        8401        8385   

   2003-12-31  2004-01-31  2004-02-29  2004-03-31  2004-04-30  2004-05-31  \
0        8379        8228        8222        8236        8191        8193   

   2004-06-30  2004-07-31  2004-08-31  2004-09-30  2004-10-31  2004-11-30  \
0        8177        8163        8105        8071        8063        8021   

   2004-12-31  2005-01-31  2005-02-28  2005-03-31  2005-04-30  2005-05-31  \
0        8009        7766        7741        7740        7729        7722   

   2005-06-30  2005-07-31  2005-08-31  2005-09-30  2005-10-31  2005-11-30  \
0        7709        7704        7658        7647        7640        7629   

   2005-12-31  2006-01-31  2006-02-28  2006-03-31  2006-04-30  2006-05-31  \
0        7604        7450        7440        7438        7432        7411   

   2006-06-30  2006-07-31  2006-08-31  2006-09-30  2006-10-31  2006-11-30  \
0        7400        7387        7355        7334        7345        7346   

   2006-12-31  2007-01-31  2007-02-28  2007-03-31  2007-04-30  2007-05-31  \
0        7303        7087        7079        7074        7069        7058   

   2007-06-30  2007-07-31  2007-08-31  2007-09-30  2007-10-31  2007-11-30  \
0        7044        7035        6950        6926        6902        6886   

   2007-12-31  2008-01-31  2008-02-29  2008-03-31  2008-04-30  2008-05-31  \
0        6873        6706        6696        6699        6708        6700   

   2008-06-30  2008-07-31  2008-08-31  2008-09-30  2008-10-31  2008-11-30  \
0        6671        6653        6633        6609        6514        6487   

   2008-12-31  2009-01-31  2009-02-28  2009-03-31  2009-04-30  2009-05-31  \
0        6475        6013        5916        5895        5883        5871   

   2009-06-30  2009-07-31  2009-08-31  2009-09-30  2009-10-31  2009-11-30  \
0        5862        5854        5850        5839        5833        5831   

   2009-12-31  2010-01-31  2010-02-28  2010-03-31  2010-04-30  2010-05-31  \
0        5819        5627        5619        5615        5610        5607   

   2010-06-30  2010-07-31  2010-08-31  2010-09-30  2010-10-31  2010-11-30  \
0        5590        5576        5562        5553        5547        5542   

   2010-12-31  2011-01-31  2011-02-28  2011-03-31  2011-04-30  2011-05-31  \
0        5526        5396        5396        5389        5387        5388   

   2011-06-30  2011-07-31  2011-08-31  2011-09-30  2011-10-31  2011-11-30  \
0        5384        5377        5368        5372        5345        5333   

   2011-12-31  2012-01-31  2012-02-29  2012-03-31  2012-04-30  2012-05-31  \
0        5312        5024        5010        4998        4994        4978   

   2012-06-30  2012-07-31  2012-08-31  2012-09-30  2012-10-31  2012-11-30  \
0        4966        4952        4924        4896

- The date columns have 10478 missing values each metro column has 14 missing values while city has 2.

- The missing values in the date columns will be filled through interpolation.
- The missing values in the metro column will be replaced with 'missing' as this may be a useful indicator.
- The missing calues in city column will be droped since it does not cover a higher percentage.

In [9]:
# imputing the missing values by replacing them with 'missing'
df.Metro.fillna('missing', inplace=True)

In [10]:
#drop missing values in the 'City' column
df.dropna(subset=['City'], inplace=True)

In [11]:
# interpolate missing values on date columns
df.interpolate(inplace=True)

In [12]:
# checking to see if missing values have been replaced
print(f'The data has {df.isna().sum().sum()} missing values')

The data has 0 missing values


### Consistency
For the data to be constistent, I need to resolve any inconsistencies by checking for duplicate values in our data.

In [13]:
# checking for duplicates
print(f'The data has {df.duplicated().sum()} duplicates')

The data has 0 duplicates


- The data has no duplicate values

### Validity

For our data to be valid, I have to verify that every column is accurate and appropriate for this analysis and remove those that are invalid.

In [14]:
#validity checks
df.head(2)

RegionID  SizeRank RegionName    RegionType StateName State       City  \
0    112345         0   Maryvale  neighborhood        AZ    AZ    Phoenix   
1    192689         1   Paradise  neighborhood        NV    NV  Las Vegas   

                              Metro       CountyName     2000-01-31  \
0         Phoenix-Mesa-Chandler, AZ  Maricopa County   67383.031641   
1  Las Vegas-Henderson-Paradise, NV     Clark County  133640.057992   

      2000-02-29     2000-03-31     2000-04-30     2000-05-31     2000-06-30  \
0   67594.528159   67876.303176   68492.532408   69174.305303   69800.457539   
1  133601.887385  133821.864483  134309.989801  134979.911695  135566.591666   

      2000-07-31     2000-08-31     2000-09-30     2000-10-31     2000-11-30  \
0   70421.136885   71051.780181   71735.008552   72362.307816   72964.869522   
1  136017.346152  136437.263631  136928.064722  137507.434436  138191.412042   

      2000-12-31     2001-01-31     2001-02-28     2001-03-31     2001-04-30  \
0   73552.527578   74153.196631   74758.925790   75378.089768   75964.876243   
1  138867.568883  139559.171348  140317.974418  141181.008999  142028.738895   

      2001-05-31     2001-06-30     2001-07-31     2001-08-31     2001-09-30  \
0   76512.540459   76998.708325   77444.604032   77861.418771   78335.805428   
1  142931.647435  143751.729034  144588.099658  145428.844650  146278.668730   

      2001-10-31     2001-11-30     2001-12-31     2002-01-31     2002-02-28  \
0   78852.272025   79407.318942   79886.223943   80287.318124   80673.299747   
1  147120.479905  147840.723518  148614.526410  149361.048529  149999.764934   

      2002-03-31     2002-04-30     2002-05-31     2002-06-30     2002-07-31  \
0   81090.978139   81511.538587   81871.363367   82228.883231   82616.778770   
1  150695.831954  151465.334047  152343.641078  153018.118424  153584.245045   

      2002-08-31     2002-09-30     2002-10-31     2002-11-30    2002-12-31  \
0   83066.909712   83471.328939   83856.974634   84195.961031   84542.84774   
1  154057.260204  154489.023461  155125.848135  156001.386636  157072.91036   

      2003-01-31     2003-02-28     2003-03-31     2003-04-30     2003-05-31  \
0   84863.049342   85184.362803   85455.290841   85794.777316   86151.639288   
1  158165.641579  159273.550003  160433.932513  161571.704926  162855.460853   

      2003-06-30     2003-07-31     2003-08-31     2003-09-30     2003-10-31  \
0   86554.870088   86935.835967   87348.756099   87862.947077   88440.516765   
1  164117.477424  165532.353597  167051.965721  168901.393777  170951.544483   

      2003-11-30     2003-12-31     2004-01-31     2004-02-29     2004-03-31  \
0   88984.482664   89432.937004   89793.043871   90235.353959   90811.466901   
1  173461.712856  176082.102702  179070.974746  182017.359314  186351.724060   

      2004-04-30     2004-05-31     2004-06-30     2004-07-31     2004-08-31  \
0   91446.641385   92185.582369   92922.032398   93992.194018   95249.242127   
1  190973.694957  196930.281540  203084.656576  211159.335478  218702.409333   

      2004-09-30     2004-10-31     2004-11-30     2004-12-31     2005-01-31  \
0   96483.444923   97693.450019   98879.462203  100402.196317  102133.562258   
1  227310.557844  233809.568726  239668.275284  243857.693634  247463.386535   

      2005-02-28     2005-03-31     2005-04-30     2005-05-31     2005-06-30  \
0  104359.759393  106590.101769  109034.773693  111640.020871  115117.796388   
1  252429.105315  257462.144898  264363.136503  269110.506250  272672.640096   

      2005-07-31     2005-08-31     2005-09-30     2005-10-31     2005-11-30  \
0  119087.454984  123056.332610  127285.319061  131123.567309  134998.558029   
1  274470.719301  277020.286129  279122.445247  280776.841322  281733.400704   

      2005-12-31     2006-01-31     2006-02-28     2006-03-31     2006-04-30  \
0  137640.827448  139268.394440  141652.773958  144685.064411  148250.083680   
1  282561.359675  283562.4

In [15]:
#droping statename since we also have state column with the same values
df.drop(columns=['StateName'],inplace=True)

### Uniformity
- For our data to be uniform, I have to verify that every column is correct and convert them to there appropriate data type.

In [16]:
#checking for data types
df.dtypes

RegionID        int64
SizeRank        int64
RegionName     object
RegionType     object
State          object
               ...   
2023-10-31    float64
2023-11-30    float64
2023-12-31    float64
2024-01-31    float64
2024-02-29    float64
Length: 298, dtype: object

### Data Preprocessing

#### Feature engineering

To improve our analysis, we will  create new features for more analytical value.

### Return on Investment (ROI)

ROI is a measure of returns expected from investments.

### Coefficient of variation (CV)

CV is a measure of the dispersion of data points around the mean and represents the ratio of the standard deviation to the mean. It allows investors to determine how much volatility, or risk, is assumed in comparison to the amount of return expected from investments.

In [17]:
#calculating and creating ROI column
df['ROI'] = (df['2024-02-29']/ df['2000-01-31'])-1


In [18]:
# Calculate mean and standard deviation along the columns axis (axis=1) for each row
mean_values = df.iloc[:, 9:].mean(axis=1)  
std_values = df.iloc[:, 9:].std(axis=1)

# Calculate coefficient of variation (CV) using the formula
cv_values = (std_values / mean_values) * 100

# Add the CV column to the DataFrame
df['CV'] = cv_values

### Reshape from wide to long format
To facilitate time series analysis, the dataframe will be transformed from a wide view to a long view format. This transformation involves restructuring the dataframe to have a single column for the Date and another column for the corresponding values. The Date column will be set as the index to establish the temporal order of the data points.

By converting the dataframe to a long view format and indexing it by the Date, the data will be organized in a sequential manner, enabling the application of time series analysis techniques such as trend analysis, seasonality detection, and forecasting. This transformation allows for a more comprehensive exploration of the temporal patterns and trends within the dataset, enhancing the accuracy and effectiveness of the subsequent time series modeling and predictions.

In [19]:
# Create a copy of the dataset to convert into long view while preserving df as a wide view for EDA
new_df = df.copy()

In [20]:
# creating a function that changes the dataframe structure from wide view to long view

def melt_df(data):
    melted = pd.melt(data, id_vars=['RegionID','RegionName','RegionType','City', 'State', 'Metro', 'CountyName', 'SizeRank',
                                  'ROI', 'CV' ], var_name='Date')
    melted['Date'] = pd.to_datetime(melted['Date'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted

In [21]:
new_df = melt_df(new_df)

C:\Users\odhiambo rodgers bon\AppData\Local\Temp\ipykernel_5916\2810894411.py:6: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  melted['Date'] = pd.to_datetime(melted['Date'], infer_datetime_format=True)


#### Convert  data to date time and indexing
Working with time series data in Python, having dates (or datetimes) in the index can be very helpful, especially if they are of DatetimeIndex type as the index allows for intuitive and efficient time-based indexing and slicing operations. It enables easy access to specific time periods, such as a particular day, month, or year. Additionally, the DatetimeIndex provides convenient methods for resampling, time shifting, and frequency conversion.

In [22]:
# Converting the date data type into date time and indexing
new_df['Date']=pd.to_datetime(new_df['Date'],format='%m/%y')

# Set the 'Date' column as index
new_df.set_index('Date',inplace=True)

In [23]:
# renaming the column value to Homevalue
new_df.rename(columns = {"value" : "Homevalue"}, inplace=True)

In [24]:
#displaying the final data
new_df.head()

RegionID         RegionName    RegionType         City State  \
Date                                                                       
2000-01-31    112345           Maryvale  neighborhood      Phoenix    AZ   
2000-01-31    192689           Paradise  neighborhood    Las Vegas    NV   
2000-01-31    270958    Upper West Side  neighborhood     New York    NY   
2000-01-31    270957    Upper East Side  neighborhood     New York    NY   
2000-01-31    118208  South Los Angeles  neighborhood  Los Angeles    CA   

                                            Metro          CountyName  \
Date                                                                    
2000-01-31              Phoenix-Mesa-Chandler, AZ     Maricopa County   
2000-01-31       Las Vegas-Henderson-Paradise, NV        Clark County   
2000-01-31  New York-Newark-Jersey City, NY-NJ-PA     New York County   
2000-01-31  New York-Newark-Jersey City, NY-NJ-PA     New York County   
2000-01-31     Los Angeles-Long Beach-Anaheim, CA  Los Angeles County   

            SizeRank       ROI         CV      Homevalue  
Date                                                      
2000-01-31         0  3.982033  59.394305   67383.031641  
2000-01-31         1  1.839812  38.067356  133640.057992  
2000-01-31         2  2.089694  29.376597  394692.554604  
2000-01-31         3  0.875548  19.879574  641748.074378  
2000-01-31         4  4.173361  48.362609  130878.148719

### Exploratory Data Analysis

In [26]:
plot_most_popular_categorical(new_df, "City")

NameError: name 'plot_most_popular_categorical' is not defined